src/utils/indexer.hpp

```diff
@@ -0,0 +1,102 @@
+//========================================================================================
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef UTILS_INDEXER_HPP_
+#define UTILS_INDEXER_HPP_
+
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+namespace parthenon {
+
+template <class... Ts>
+struct Indexer {
+  Indexer() : N{}, start{}, _size{} {};
+
+  explicit Indexer(std::pair<Ts, Ts>... Ns)
+      : N{GetFactors(std::make_tuple((Ns.second - Ns.first + 1)...),
+                     std::make_index_sequence<sizeof...(Ts)>())},
+        start{Ns.first...}, end{Ns.second...}, _size(((Ns.second - Ns.first + 1) * ...)) {
+  }
+
+  Indexer(Indexer &&) = default;
+  Indexer(const Indexer &) = default;
+  Indexer &operator=(const Indexer &) = default;
+  Indexer &operator=(Indexer &&) = default;
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  std::size_t size() const { return _size; }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  std::tuple<Ts...> operator()(int idx) const {
+    return GetIndicesImpl(idx, std::make_index_sequence<sizeof...(Ts)>());
+  }
+
+  template <std::size_t I>
+  KOKKOS_FORCEINLINE_FUNCTION auto StartIdx() const {
+    return std::get<I>(start);
+  }
+
+  template <std::size_t I>
+  KOKKOS_FORCEINLINE_FUNCTION auto EndIdx() const {
+    return std::get<I>(end);
+  }
+
+  static const constexpr std::size_t rank = sizeof...(Ts);
+
+ private:
+  template <std::size_t... Is>
+  KOKKOS_FORCEINLINE_FUNCTION std::tuple<Ts...>
+  GetIndicesImpl(int idx, std::index_sequence<Is...>) const {
+    std::tuple<Ts...> idxs;
+    (
+        [&] {
+          std::get<Is>(idxs) = idx / std::get<Is>(N);
+          idx -= std::get<Is>(idxs) * std::get<Is>(N);
+          std::get<Is>(idxs) += std::get<Is>(start);
+        }(),
+        ...);
+    return idxs;
+  }
```

Comment:
I like it

Reply:
yeah, I was pretty proud of this code. It should produce the same assembly if you had written out the indexing by hand. Sadly, it may not be usable because the tuple copy constructor doesn't seem to work on device. Maybe the compiler option `--expt-relaxed-constexpr` would fix this for CUDA but it also seems to not work for HIP.

Reply:
It looks like the solution to the shortcomings of `std::tuple` on device can be circumvented by just switching to `std::array` so we can use the generic implementation.

Reply:
FWIW there's also a `Kokkos::Array`